File: src/fwt/js/FwtGraphicsPath.js
//
// Copyright (c) 2014, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   24 Feb 14  Brian Frank  Creation
//

/**
 * FwtGraphicsPath implements gfx::GraphicsPath using HTML5 canvas.
 */

fan.fwt.FwtGraphicsPath = fan.sys.Obj.$extend(fan.sys.Obj);
fan.fwt.FwtGraphicsPath.prototype.$ctor = function() {}

fan.fwt.FwtGraphicsPath.prototype.$typeof = function()
{
  return fan.fwt.FwtGraphicsPath.$type;
}

// canvas context
fan.fwt.FwtGraphicsPath.prototype.cx = null;

// This draw()
fan.fwt.FwtGraphicsPath.prototype.draw = function()
{
  this.cx.stroke();
  return this;
}

// This fill()
fan.fwt.FwtGraphicsPath.prototype.fill = function()
{
  this.cx.fill();
  return this;
}

// This clip()
fan.fwt.FwtGraphicsPath.prototype.clip = function()
{
  this.cx.clip();
  return this;
}

// This moveTo(Int x, Int y)
fan.fwt.FwtGraphicsPath.prototype.moveTo = function(x, y)
{
  this.cx.moveTo(x, y);
  return this;
}

// This lineTo(Int x, Int y)
fan.fwt.FwtGraphicsPath.prototype.lineTo = function(x, y)
{
  this.cx.lineTo(x, y);
  return this;
}

// This curveTo(Int cp1x, Int cp1y, Int cp2x, Int cp2y, Int x, Int y)
fan.fwt.FwtGraphicsPath.prototype.curveTo = function(cp1x, cp1y, cp2x, cp2y, x, y)
{
  this.cx.bezierCurveTo(cp1x+0.5, cp1y+0.5, cp2x+0.5, cp2y+0.5, x+0.5, y+0.5);
  return this;
}

// This close()
fan.fwt.FwtGraphicsPath.prototype.close = function()
{
  this.cx.closePath();
  return this;
}

File: src/fwt/java/FwtGraphicsPath.java
//
// Copyright (c) 2014, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   24 Jun 14  Brian Frank  Creation
//
package fan.fwt;

import fan.sys.UnsupportedErr;
import fan.gfx.*;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.GC;
import org.eclipse.swt.graphics.Path;

public class FwtGraphicsPath implements GraphicsPath
{
  public FwtGraphicsPath(FwtGraphics g)
  {
    this.g = g;
    this.path = new Path(g.gc.getDevice());
  }

  public GraphicsPath draw()
  {
    g.gc.drawPath(path);
    path.dispose();
    return this;
  }

  public GraphicsPath fill()
  {
    g.gc.fillPath(path);
    path.dispose();
    return this;
  }

  public GraphicsPath clip()
  {
    path.dispose();
    throw UnsupportedErr.make("SWT GraphicPath.clip");
  }

  public GraphicsPath moveTo(long x, long y)
  {
    path.moveTo((int)x, (int)y);
    return this;
  }

  public GraphicsPath lineTo(long x, long y)
  {
    path.lineTo((int)x, (int)y);
    return this;
  }

  public GraphicsPath curveTo(long cp1x, long cp1y, long cp2x, long cp2y, long x, long y)
  {
    path.cubicTo((int)cp1x, (int)cp1y, (int)cp2x, (int)cp2y, (int)x, (int)y);
    return this;
  }

  public GraphicsPath close()
  {
    path.close();
    return this;
  }

  final FwtGraphics g;
  final Path path;

}